storage/database/src/key.rs

```diff
@@ -1,54 +1,180 @@
 //! Database key abstraction; `trait Key`.
 
 //---------------------------------------------------------------------------------------------------- Import
-use std::cmp::Ordering;
+use std::{cmp::Ordering, fmt::Debug};
 
-use crate::storable::Storable;
+use crate::{storable::Storable, StorableBytes, StorableStr, StorableVec};
 
 //---------------------------------------------------------------------------------------------------- Table
 /// Database [`Table`](crate::table::Table) key metadata.
 ///
 /// Purely compile time information for database table keys.
-//
-// FIXME: this doesn't need to exist right now but
-// may be used if we implement getting values using ranges.
+///
+/// ## Comparison
+/// There are 2 differences between [`Key`] and [`Storable`]:
+/// 1. [`Key`] must be [`Sized`]
+/// 2. [`Key`] represents a [`Storable`] type that defines a comparison function
+///
+/// The database backends will use [`Key::KEY_COMPARE`]
+/// to sort the keys within database tables.
+///
+/// [`Key::KEY_COMPARE`] is pre-implemented as a straight byte comparison.
+///
+/// This default is overridden for numbers, which use a number comparison.
+/// For example, [`u64`] keys are sorted as `{0, 1, 2 ... 999_998, 999_999, 1_000_000}`.
+///
+/// If you would like to re-define this for number types, consider;
+/// 1. Creating a wrapper type around primitives like a `struct SortU8(pub u8)`
+/// 2. Implement [`Key`] on that wrapper
+/// 3. Define a custom [`Key::KEY_COMPARE`]
+// FIXME:
+// implement getting values using ranges.
 // <https://github.com/Cuprate/cuprate/pull/117#discussion_r1589378104>
-pub trait Key: Storable + Sized {
-    /// The primary key type.
-    type Primary: Storable;
-
+pub trait Key: Storable + Sized + Ord {
```

Comment:
`Key` must now be `Ord` because `heed`'s `create_db` may call:
- `heed::Comparator::compare` which may call...
- `KeyCompare::as_compare_fn` which may call...
- `Ord::cmp()` for value comparison

Thankfully this is already the case for all our keys but is this going to be a problem in the future? Is there a clearer way to set these bounds?

Reply:
I don't think so, it kinda makes sense for `Key` to require `Ord`